107-tee-boot-attestation.md

```diff
@@ -0,0 +1,360 @@
+ * SVSM sends all attestation information to TEE handler.
+ * TEE handler uses information to verify the boot environment.
+ * If boot attestation is successful, TEE handler reads the agent's TEE secret
+   from the registrar DB and sends to CVM. If boot attestation fails, TEE
+   handler sends error message.
+ * SVSM reads response from TEE handler. If successful, SVSM uses the TEE secret
+   to unlock some state and begin booting guest OS, with a keylime agent
+   eventually initializing and integrity monitoring continuing as normal.
+ * SVSM will extend system's TPM state to account for TEE boot attestation.
+   *This is intentially vague at the moment, as the exact design is still under
+   ongoing discussion*.
+ * Agent initializes and begins running as normal. Relevant additions are made
+   to verifier to check if TPM measurement includes the data returned from TEE
+   handler. *This is also still under ongoing discussion*.
+
+### Notes/Constraints/Caveats (optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above.
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+ * TODO
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal and how do we mitigate.  Think broadly.
+For example, consider both security and how this will impact the larger
+enhancement ecosystem.
+
+How will security be reviewed and by whom?
+-->
+
+The TEE handler will contain a very sensitive secret for unlocking some state
+on the CVM. This secret must be properly protected and should only be made
+available to a guest that has successfully attested.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+The keylime verifier will implement a new handler at `/v_/verify/tee` which can
+be accessed with a `POST` method. The request body is a JSON object with the
+following properties:
+
+- `tee` (_string_) - TEE architecture that the `evidence` will be formatted in.
+- `evidence` (_JSON object_) - TEE evidence. This will be deserialized
+  differently depending on the TEE architecture specified in `tee`.
+- `pubkey_pem` (_string_) - PEM-encoded public key to encrypt the TEE secret
+  with upon a successful attestation.
+
+The RESTful parameters of this handler would also need to include an agent UUID,
+as TEE secrets are specific to each agent.
+
+Upon a `POST` request to the TEE handler, the handler will deserialize the
+request body from JSON to fetch the parameters listed above. The handler will
+read the `tee` parameter and deserialize the `evidence` parameter accordingly.
```

Comment:
Presumably the tenant tool will be used to set how to recognize 'whatever is expected to be good as evidence'.

Reply:
Can you elaborate a bit more? There would probably need to be a "standard", agreed upon format that each TEE architecture's JSON evidence will need to follow.

Reply:
A user will probably need to be able to put 'the very sensitive secret' (or something related to 'it') into the DB using the keylime tenant tool so that he verifier can then either compare the 'very sensitive secret' produced by the TEE or do some computation (with something relate to 'it') to be able to compare it.